### ISSUE TITLE
Re-evaluate contract of StatementList::toArray

### ISSUE THREAD

**Issue by JeroenDeDauw** [wmde/WikibaseDataModel#446]
Right now the docs state

    @return Statement[] Numerically indexed (non-sparse) array.

If we add the ability to remove elements from the list, the sparseness will have to be death with somewhere. Which we probably do not want to be in this class, since then we'll need to pay the performance penalty for each `toArray` call, even if most users don't care about the keys.

**Comment by thiemowmde** [comment]
Whats the problem with that? Currently the class can give this guarantee and I think it should  continue to do so.

**Comment by thiemowmde** [comment]
Ok, what we can do:
* Make `toArray` return an array of Statement objects, indexed by GUID.
* Calling code can do `array_values` if required.

**Comment by JeroenDeDauw** [comment]
The idea behind the toArray method is to just return the array the class wraps, so one can as cheaply as possible construct another collection type. It also allows doing

```php
foreach ( $statementList->toArray() as $statement ) {
}
```

Which then makes tools understand that $statement is a `Statement` without adding a specific tag (at the cost of a cheap method call once for the entire loop).

So I think it's indeed better if the caller is responsible for doing the `array_values` operation in case they care.

**Comment by thiemowmde** [comment]
Wait a sec. I wonder:
1. The new `StatementList::getIndexByGuid` is already not very useful, because accessing Statements by index is nothing we do often. And it becomes useless the moment we are dropping this guarantee. Should we remove it?
2. Indexing the internal $statements array by GUID is not really possible because a Statement is not guaranteed to have a GUID set.

**Comment by JeroenDeDauw** [comment]
> The discussion in #453 indicates that the contract of this method should not be changed.

How?

**Comment by thiemowmde** [comment]
#453 answers the question how it should **not** change. The remaining question is how you want to change it and why?

**Comment by JeroenDeDauw** [comment]
The reason why I think the current description is problematic is described in the issue here. You've now stated twice that 453 answers describes you think it should not change. Again, how does it describe this? I read through it and am not seeing what the reason is.

**Comment by thiemowmde** [comment]
I can not see how a sparse array can make it into the class, so I wonder why we are discussing this? Why should we **drop** the guarantee and not replace it with an other useful one? Indexing by GUID is out, which is fine. Is there an other option that makes sense?

**Comment by JeroenDeDauw** [comment]
> I can not see how a sparse array can make it into the class

It cannot now. I'm thinking about the future. Please read my first comment once again.

**Comment by thiemowmde** [comment]
> Please read my first comment once again.

Why? Did you edited it?

> If we add the ability to remove elements […]

How is that an argument? We can give the guarantee and do so. What's the benefit of dropping it? Do you plan to add a remove method? And if you do so, why does it need to be cheap? Such a method isn't used very often. It can easily reindex the internal array with fast PHP function calls. How can the hypothetical performance of non-existing methods be an argument to drop a useful feature?

**Comment by thiemowmde** [comment]
@JeroenDeDauw wrote in https://github.com/wmde/WikibaseDataModel/pull/466#issuecomment-96851750:
> -2. @thiemowmde I'm not sure what you do not understand about the change. The problem was outlined and other people seem to understand it.

I'm repeating myself: What problem? How is a hypothetical performance issue in methods that do not exist a problem that needs to be solved now? This introduces other problems that are, in my opinion, not outweigh by the advantage.

**Comment by thiemowmde** [comment]
I did some micro benchmarks:
* `array_splice` (removes elements from an array with reindexing) on 10,000 arrays with 100 elements each needs ~70ms.
* `unset` (removes an element without reindexing) on the same arrays needs ~10ms.
* `array_values` to reindex the 10000 sparse arrays needs ~80ms.

Without the guarantee certain callers of `toArray` (`foreach` being one of a few exceptions) **must** reindex the array with an additional `array_values` because they just don't know if the array is sparse. But in almost all cases it's not and the additional `array_values` is doing nothing but wasting time. Why should it be sparse? Currently it just **can't** be sparse. Even with the proposed remove method the array will **not** be sparse in almost all cases.

Furthermore, **not** giving the guarantee is error-prone because users will forget and be surprised when they get a sparse array. This will lead to bugs that are usually not covered by tests and will first show up in production.

With the guarantee all methods that are manipulating the internal array must take care of the indexes. That would be a straight `array_splice` in the proposed remove method. Done. This needs a little time when removing an element (how often does this happen?) but saves time whenever `toArray` is called and an `array_values` can be saved.

**Comment by JeroenDeDauw** [comment]
So you propose to force all removal methods to reindex the array, also saying that this is fine performance wise? If it is fine performance wise, then that approach is fine with me. Doing a reindex on every remove operation seems quite inefficient though. Anyway, it is now clear you understand my concern. While I'm still dubious about this trade-off, I'm not going to block you from making it if you take responsibility for it.

**Comment by thiemowmde** [comment]
Actually I propose to never add a remove method, which makes all discussion obsolete. In my opinion this is just not needed and also violates the "does not provide complex modification functionality" contract of the class.

**Comment by thiemowmde** [comment]
Actually I propose to never add a remove method, which makes all discussion obsolete. I suggested this once (in #175, now tracked in [T78299](https://phabricator.wikimedia.org/T78299)) but when I worked on phasing out `Claims` (most notably https://gerrit.wikimedia.org/r/#/c/204511/, which is still not reviewed and merged for a reason I do not understand) it turned out that a remove method is just not needed. As suggested in https://github.com/wmde/WikibaseDataModel/pull/455#issuecomment-94516582 I agree that a remove method violates the "does not provide complex modification functionality" contract of the class.

**Comment by JeroenDeDauw** [comment]
We can also do that yes - decide that this class will not get any remove methods.

We should first make sure there are no obvious problems this will cause. To me it seems that we likely do have use for a `removeStatementByGuid` method. Yes, you can create a `StatementByGuidMap` and us its removal methods, but that makes no sense if you just want to remove a single Statement by guid.

Removing a Statement by guid does not violate the contract of the class. If it did, we'd not be having this discussion. And save me the arguments that this falls under "complex modification functionality". It certainly does not match what I intended to describe when writing that. So such bickering about terminology is completely besides the point.

**Comment by Benestar** [comment]
@thiemowmde Currently it is not possible to remove a `Statement` from a `StatementListProvider` so we either need a `remove` method in `StatementList` or a setter in `StatementListProvider`.

**Comment by thiemowmde** [comment]
We need neither nor, proven by https://gerrit.wikimedia.org/r/#/c/204511/. And even if we ever implement a remove method (which still is purely theoretical since I have never seen a pull request nor urgent ticket nor the actual need for it) this still does not mean we must drop the guarantee given on the toArray method. Especially not **before** the method is actually implemented.

TL;DR:
* A "provider" can not have a setter.
* StatementList is immutable.
* Making it mutable does not mean the guarantee must be dropped.

**Comment by Benestar** [comment]
> StatementList is immutable

W00t??? Having a `addStatement` method clearly indicates that this class *is* mutable.

**Comment by thiemowmde** [comment]
Not relevant for the toArray discussion.

**Comment by Benestar** [comment]
Hmm, you mentioned all this points and now say they are not relevant?

**Comment by thiemowmde** [comment]
Yes, since you asked, but I'm not going to continue discussing details that have it's own thread. Moreover, I will avoid continuing discussions here on GitHub. Please use Phabricator.

**Comment by Benestar** [comment]
Ok, I will create a task on Phabricator. However, I'm not the person who startet this thread on Github.
